sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -51,6 +51,11 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
 
   def reorder(plan: LogicalPlan, output: AttributeSet): LogicalPlan = {
     val (items, conditions) = extractInnerJoins(plan)
+    // Find the star schema joins. Currently, it returns the star join with the largest
+    // fact table. In the future, it can return more than one star join (e.g. F1-D1-D2
+    // and F2-D3-D4).
+    val starJoinPlans = StarSchemaDetection(conf).findStarJoins(items, conditions.toSeq)
```

Comment:
@ron8hu One way to overcome the search space problem of a dynamic programming join enumeration is to support various levels of optimization algorithms. For example, the lowest level would employ a “greedy” algorithm that would choose the “cheapest” join that can be made over the remaining sub-plans in each iteration. Then, we move up to higher levels of optimizations such as dynamic programming join enumeration with various heuristics. Some good heuristics that can be applied are: push Cartesian products planning towards the end; produce only left-deep tree; apply star schema heuristics, etc. The next level of optimization would be full, maximal optimization that considers all possible join orders and spares no effort… Catalyst/CBO can start with one level optimization, and automatically drop to “greedy” if it exceeds the allocated search space. We can also give control to the user by externalizing these levels of optimizations. If their query compile time is too high, they can lower the level.